Create rule S7011: Detect Azure EventGrid SAS (#4055)
SonarSource/rspec: rules/S7011/secrets/rule.adoc

include::../../../shared_content/secrets/description.adoc[]

== Why is this an issue?

include::../../../shared_content/secrets/rationale.adoc[]

=== What is the potential impact?

Access keys are used to authenticate to Azure Event Grid resources and can be
employed for event publication to the resource.

Shared Access Secrets (SAS) are generated from access keys and used to
authenticate specific resources with particular permissions within a defined
time frame. +
Despite their risk reduction on the overall infrastructure, hardcoding them

> exposes them to exploitation to access the resource. This is a risk, unless the

Reviewer: > Despite their risk reduction on the overall infrastructure

I am not sure why hardcoding them reduces the risk.

> hardcoding them exposes them to exploitation to access the resource

That doesn't sound like correct English.

> Despite their risk reduction on the overall infrastructure, hardcoding them
> exposes them to exploitation to access the resource. This is a risk, unless the
> whole resource is free of sensitive data.

Reviewer 2: ```suggestion
Using Shared Access Signatures is always better than using access keys,
because they reduce risk thanks to their scope and time limitation.
However, hardcoding secret tokens is always a bad practice, and still
exposes the resources to attacks. +
This is a risk unless the whole resource is free of sensitive data.
```

> *Please note* that the Microsoft docs can use different terminologies for these

Reviewer: Feels a bit arbitrary to just make the first two words bold.

Reviewer 2: ```suggestion
Note that the Microsoft docs can use different terminologies for these
```

// When using predefined content, search for any required variables to be set and include them in this file.
// Not adding them will not trigger warnings.

include::../../../shared_content/secrets/impact/non_repudiation.adoc[]

include::../../../shared_content/secrets/impact/data_compromise.adoc[]

include::../../../shared_content/secrets/impact/financial_loss.adoc[]

== How to fix it

include::../../../shared_content/secrets/fix/revoke.adoc[]

include::../../../shared_content/secrets/fix/vault.adoc[]

On top of that, use the Azure SDK as much as possible and their diverse
credential objects.

=== Code examples

==== Noncompliant code example

For Shared Access Secrets:

[source,python]
----
topic_endpoint = "https://example.westeurope-1.eventgrid.azure.net/api/events"

headers = {
    "aeg-sas-token": "r=https%3a%2f%2fexample.westeurope-1.eventgrid.azure.net%2fapi%2fevents&e=1%2f1%2f0001%2012%3a02%3a52%20AM&s=gdcIr5dXR4dC3sNyK4Qree4XogaRma3YdtH7%2boCxSQo%3d", # Noncompliant
    "Content-Type": "application/json"
}

event_data = [{ ... }]

response = requests.post(topic_endpoint, headers=headers, data=json.dumps(event_data))
----

> For Access Keys:

Reviewer: I think one example would have been enough, considering the second one just has a slightly different variable name. Any reason why there is no compliant solution?

Reviewer 2: I added the two examples to make sure the users understand which one is which. I added no compliant solution since the goal is to stop hardcoding them. Do you think I should add something, like an example of the Azure SDK?

Reviewer: I think there is a confusion of terminology? Our recommended solution for hardcoded secrets is always not to hardcode them. That's why the default suggestion is to pass it to the app through environment variables.

Currently, the suggested fix in this RSPEC is to use a secret vault. If that is not the actual suggested fix, then I guess it has to be replaced.

two tokens:

* Access Keys are also called SAS keys
* Shared Access Secrets are also called SAS tokens

Below are some real-world scenarios that illustrate some impacts of an attacker
exploiting these types of secrets.

// Set value that can be used to refer to the type of secret in, for example:
// "An attacker can use this {secret_type} to ..."
:secret_type: secret

// Where possible, use predefined content for common impacts. This content can
// be found in the folder "shared_content/secrets/impact".
// When using predefined content, search for any required variables to be set and include them in this file.
// Not adding them will not trigger warnings.

include::../../../shared_content/secrets/impact/non_repudiation.adoc[]

include::../../../shared_content/secrets/impact/data_compromise.adoc[]

include::../../../shared_content/secrets/impact/financial_loss.adoc[]

== How to fix it

include::../../../shared_content/secrets/fix/revoke.adoc[]

include::../../../shared_content/secrets/fix/vault.adoc[]

On top of that, use the Azure SDK as much as possible and their diverse
credential objects.

=== Code examples

==== Noncompliant code example

> For Shared Access Secrets:
> 
> [source,python]
> ----
> topic_endpoint = "https://example.westeurope-1.eventgrid.azure.net/api/events"
> 
> headers = {
>     "aeg-sas-token": "r=https%3a%2f%2fexample.westeurope-1.eventgrid.azure.net%2fapi%2fevents&e=1%2f1%2f0001%2012%3a02%3a52%20AM&s=gdcIr5dXR4dC3sNyK4Qree4XogaRma3YdtH7%2boCxSQo%3d", # Noncompliant
>     "Content-Type": "application/json"
> }
> 
> event_data = [{ ... }]
> 
> response = requests.post(topic_endpoint, headers=headers, data=json.dumps(event_data))
> ----
> 
> For Access Keys:
> 
> [source,python]
> ----
> topic_endpoint = "https://example.westeurope-1.eventgrid.azure.net/api/events"
> 
> headers = {
>     "aeg-sas-key": "vobPdVO3kYZKwbnil2IsmyrR/gTYenN4vAZEGHUfhj8=", # Noncompliant
>     "Content-Type": "application/json"
> }
> 
> event_data = [{ ... }]
> 
> response = requests.post(topic_endpoint, headers=headers, data=json.dumps(event_data))
> ----

Reviewer 2: ```suggestion

:example_secret: r=https%3a%2f%2fexample.westeurope-1.eventgrid.azure.net%2fapi%2fevents&e=1%2f1%2f0001%2012%3a02%3a52%20AM&s=gdcIr5dXR4dC3sNyK4Qree4XogaRma3YdtH7%2boCxSQo%3d
:example_name: aeg-sas-token
:example_env: AEG_SAS_TOKEN

include::../../../shared_content/secrets/examples.adoc[]
```
